modules/generic-scala3/src/main/scala/pureconfig/generic/scala3/HintsAwareProductConfigReaderDerivation.scala

```diff
@@ -0,0 +1,137 @@
+package pureconfig
+package generic
+package scala3
+
+import scala.compiletime._
+import scala.compiletime.ops.int._
+import scala.deriving.Mirror
+import scala.quoted._
+
+import pureconfig.error.{ConfigReaderFailures, KeyNotFound, WrongSizeList}
+import pureconfig.generic.ProductHint.UseOrDefault
+import pureconfig.generic.derivation.Utils
+import pureconfig.generic.derivation.Utils.widen
+
+trait HintsAwareProductConfigReaderDerivation { self: HintsAwareConfigReaderDerivation =>
+  inline def deriveProductReader[A](using
+      m: Mirror.ProductOf[A],
+      ch: CoproductHint[A],
+      ph: ProductHint[A]
+  ): ConfigReader[A] =
+    inline erasedValue[A] match {
+      case _: Tuple =>
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): ConfigReader.Result[A] =
+            for {
+              listCur <- asList(cur)
+              result <- readTuple[A & Tuple, 0](listCur.list)
+            } yield result
+
+          def asList(cur: ConfigCursor) =
+            cur.asListCursor.flatMap { listCur =>
+              if (constValue[Tuple.Size[A & Tuple]] == listCur.size)
+                Right(listCur)
+              else
+                listCur.failed(
+                  WrongSizeList(constValue[Tuple.Size[A & Tuple]], listCur.size)
+                )
+            }
+        }
+
+      case _ =>
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): ConfigReader.Result[A] = {
+            val tupleSize = summonInline[ValueOf[Tuple.Size[m.MirroredElemTypes]]]
+            val defaults = ProductDerivationMacros.getDefaults[A](tupleSize.value)
+
+            for {
+              objCursor <- cur.asObjectCursor
+              labels = Utils.transformedLabels(identity)
+              actions = labels.map { label => label -> ph.from(objCursor, label) }.toMap
+              result <- readCaseClass[m.MirroredElemTypes, 0, A](objCursor, labels, actions, defaults)
+            } yield m.fromProduct(result)
+          }
+        }
+    }
+
+  private inline def readCaseClass[T <: Tuple, N <: Int, A: ProductHint: CoproductHint](
```

Comment:
From this point on we shouldn't need any implicit hints, as we already extracted actions for each key.

Reply:
We don't need the `CoproductHint`, but `ProductHint` is needed to access the `.bottom`, even with extracted actions. I can however pass it explicitly.